yt/visualization/fits_image.py

```diff
@@ -906,17 +910,20 @@ def construct_image(
             frb = data_source.to_frb(width[0], (nx, ny), height=width[1])
         else:
             frb = data_source.to_frb(width[0], (nx, ny), center=center, height=width[1])
-    elif isinstance(data_source, ParticleAxisAlignedDummyDataSource):
-        axes = axis_wcs[axis]
-        bounds = (
-            center[axes[0]] - width[0] / 2,
-            center[axes[0]] + width[0] / 2,
-            center[axes[1]] - width[1] / 2,
-            center[axes[1]] + width[1] / 2,
-        )
-        frb = ParticleImageBuffer(
-            data_source, bounds, (nx, ny), periodic=all(ds.periodicity)
-        )
+    elif isinstance(data_source, ParticleDummyDataSource):
+        if hasattr(data_source, "normal_vector"):
```

Comment:
`data_source.normal_vector` is not actually used in the following so it is a bit obscure why the branching is done depending on wether or not this attribute exis. If we're going to treat `OffAxis...DummyDataSource` classes as a protocol, I'd rather we do it more explicitly (yes it's uglier, but also much easier to understand and refactor if we ever manage to)

```suggestion
        if re.fullmatch(r"OffAxis\w+DummyDataSource", type(data_source).__name__):
```

Reply:
I'd prefer we find a less ugly way to do this.

Reply:
It's frankly better to add a comment here why we check this.

Reply:
There's way to implement this check using `isinstance` instead of `hasattr` that
* can be checked continuously
* doesn't require a (possibly outdated) comment 
* doesn't require inheritance

Namely: [`typing.Protocol`](https://docs.python.org/3/library/typing.html#typing.Protocol) (which we do have access to since we don't support Python < 3.8 any more)

For illustration, here's a minimal implementation
```python
from typing import Protocol, runtime_checkable
import numpy as np

@runtime_checkable
class OffAxisDummyDataSource:
    normal_vector: np.ndarray

class Dumdumdatasource:
    normal_vector = np.zeros(3)
    
assert isinstance(Dumdumdatasource(), OffAxisDummyDataSource)
```

I understand that this goes beyond the scope of this PR, which is why I instead suggested an (ugly) one-line refactor to make the code more expressive, but long term I'd like to start using such protocol classes to replace calls to `hasattr`; IMO they would fit nicely in the `yt._typing` namespace. Is this something we can all get behind ?

Reply:
I'd like an answer to this before I hit approve. I'm happy to put in the work for yt 4.3.0

Reply:
I'm obviously happy to discuss anything, and it does look like a better long-term solution, but needless to say we'll have to hash that out in a different forum. 

That ok?

Reply:
Of course, I just didn't want to leave unresolved threads before merging this important PR.